[WPE] Use new Pasteboard API

carlosgcampos opened the pull request:
#### a3ac190a7ddfa94d5e1acf68b7b54ff846391ede
<pre>
[WPE] Use new Pasteboard API
<a href="https://bugs.webkit.org/show_bug.cgi?id=291935">https://bugs.webkit.org/show_bug.cgi?id=291935</a>

Reviewed by Michael Catanzaro.

Switch to use the new Pasteboard API, reusing GTK implementation when
possible. This also fixes a crash with the new API when the pasteboard
is used, since it uses libwpe unconditionally. New Pasteboard API allows
to enable several features that were disabled for WPE, but they won&apos;t
work with the old API because libwpe only supports text items in clipboard.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/editing/WebContentReader.cpp:
* Source/WebCore/editing/WebContentReader.h:
* Source/WebCore/editing/glib/EditorGLib.cpp: Renamed from Source/WebCore/editing/gtk/EditorGtk.cpp.
(WebCore::Editor::pasteWithPasteboard):
(WebCore::Editor::platformCopyFont):
(WebCore::Editor::platformPasteFont):
(WebCore::elementURL):
(WebCore::getImageForElement):
(WebCore::Editor::writeImageToPasteboard):
(WebCore::Editor::writeSelectionToPasteboard):
(WebCore::Editor::webContentFromPasteboard):
* Source/WebCore/editing/glib/WebContentReaderGLib.cpp: Renamed from Source/WebCore/editing/gtk/WebContentReaderGtk.cpp.
(WebCore::WebContentReader::readFilePath):
(WebCore::WebContentReader::readHTML):
(WebCore::WebContentReader::readPlainText):
(WebCore::WebContentReader::readImage):
(WebCore::WebContentReader::readURL):
(WebCore::shouldReplaceSubresourceURL):
(WebCore::WebContentMarkupReader::readHTML):
* Source/WebCore/platform/Pasteboard.h:
* Source/WebCore/platform/PasteboardCustomData.h:
(WebCore::PasteboardCustomData::wpeType):
* Source/WebCore/platform/PasteboardStrategy.h:
* Source/WebCore/platform/PlatformPasteboard.h:
* Source/WebCore/platform/SourcesGLib.txt:
* Source/WebCore/platform/glib/SelectionData.cpp: Renamed from Source/WebCore/platform/gtk/SelectionData.cpp.
(WebCore::SelectionData::SelectionData):
(WebCore::replaceNonBreakingSpaceWithSpace):
(WebCore::SelectionData::setText):
(WebCore::SelectionData::setURIList):
(WebCore::SelectionData::setURL):
(WebCore::SelectionData::urlLabel const):
(WebCore::SelectionData::clearAllExceptFilenames):
(WebCore::SelectionData::clearAll):
* Source/WebCore/platform/glib/SelectionData.h: Renamed from Source/WebCore/platform/gtk/SelectionData.h.
(WebCore::SelectionData::text const):
(WebCore::SelectionData::hasText const):
(WebCore::SelectionData::clearText):
(WebCore::SelectionData::setMarkup):
(WebCore::SelectionData::markup const):
(WebCore::SelectionData::hasMarkup const):
(WebCore::SelectionData::clearMarkup):
(WebCore::SelectionData::url const):
(WebCore::SelectionData::hasURL const):
(WebCore::SelectionData::clearURL):
(WebCore::SelectionData::uriList const):
(WebCore::SelectionData::filenames const):
(WebCore::SelectionData::hasURIList const):
(WebCore::SelectionData::hasFilenames const):
(WebCore::SelectionData::clearURIList):
(WebCore::SelectionData::setImage):
(WebCore::SelectionData::image const):
(WebCore::SelectionData::hasImage const):
(WebCore::SelectionData::clearImage):
(WebCore::SelectionData::setCanSmartReplace):
(WebCore::SelectionData::canSmartReplace const):
(WebCore::SelectionData::addBuffer):
(WebCore::SelectionData::buffers const):
(WebCore::SelectionData::buffer):
(WebCore::SelectionData::clearBuffers):
(WebCore::SelectionData::setCustomData):
(WebCore::SelectionData::customData const):
(WebCore::SelectionData::hasCustomData const):
(WebCore::SelectionData::clearCustomData):
* Source/WebCore/platform/libwpe/PlatformPasteboardLibWPE.cpp:
(WebCore::PlatformPasteboard::changeCount const):
(WebCore::PlatformPasteboard::write):
* Source/WebCore/platform/wpe/PasteboardWPE.cpp: Added.
(WebCore::Pasteboard::createForCopyAndPaste):
(WebCore::Pasteboard::Pasteboard):
(WebCore::Pasteboard::writeString):
(WebCore::Pasteboard::writePlainText):
(WebCore::Pasteboard::write):
(WebCore::Pasteboard::writeTrustworthyWebURLsPboardType):
(WebCore::Pasteboard::clear):
(WebCore::Pasteboard::canSmartReplace):
(WebCore::Pasteboard::read):
(WebCore::Pasteboard::hasData):
(WebCore::Pasteboard::typesSafeForBindings):
(WebCore::Pasteboard::typesForLegacyUnsafeBindings):
(WebCore::Pasteboard::readOrigin):
(WebCore::Pasteboard::readString):
(WebCore::Pasteboard::readStringInCustomData):
(WebCore::Pasteboard::fileContentState):
(WebCore::Pasteboard::writeMarkup):
(WebCore::Pasteboard::writeCustomData):
(WebCore::Pasteboard::changeCount const):
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Scripts/webkit/messages.py:
(conditions_for_header):
* Source/WebKit/Shared/Pasteboard.serialization.in:
* Source/WebKit/Shared/glib/SelectionData.serialization.in: Renamed from Source/WebKit/Shared/gtk/ArgumentCodersGtk.serialization.in.
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/WebPasteboardProxy.cpp:
* Source/WebKit/UIProcess/WebPasteboardProxy.h:
* Source/WebKit/UIProcess/WebPasteboardProxy.messages.in:
* Source/WebKit/UIProcess/wpe/WebPasteboardProxyWPE.cpp: Added.
(WebKit::WebPasteboardProxy::getTypes):
(WebKit::WebPasteboardProxy::readText):
(WebKit::WebPasteboardProxy::readFilePaths):
(WebKit::WebPasteboardProxy::readBuffer):
(WebKit::WebPasteboardProxy::writeToClipboard):
(WebKit::WebPasteboardProxy::clearClipboard):
(WebKit::WebPasteboardProxy::typesSafeForDOMToReadAndWrite):
(WebKit::WebPasteboardProxy::writeCustomData):
(WebKit::WebPasteboardProxy::allPasteboardItemInfo):
(WebKit::WebPasteboardProxy::informationForItemAtIndex):
(WebKit::WebPasteboardProxy::getPasteboardItemsCount):
(WebKit::WebPasteboardProxy::readURLFromPasteboard):
(WebKit::WebPasteboardProxy::readBufferFromPasteboard):
(WebKit::WebPasteboardProxy::getPasteboardChangeCount):
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.h:

Canonical link: <a href="https://commits.webkit.org/294103@main">https://commits.webkit.org/294103@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c963b9ab51b2870d46ba4e23c47e9443d83c1a41

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100841 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20493 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10792 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105978 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51430 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102882 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20802 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28967 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76785 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33824 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103848 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15992 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91084 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57137 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/100315 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15803 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9093 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50807 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/93505 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85703 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9167 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108333 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/99449 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27959 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85748 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28322 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87286 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85291 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30004 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7733 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21973 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16402 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27894 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33152 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/123075 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27705 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34298 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31023 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29263 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/wpe/PasteboardWPE.cpp:
```diff
@@ -0,0 +1,262 @@
+        return;
+    }
+
+    if (reader.shouldReadBuffer("image/png"_s)) {
+        if (auto buffer = readBuffer(index, "image/png"_s))
+            reader.readBuffer({ }, { }, buffer.releaseNonNull());
+    }
```
Why is image/png hardcoded here?

carlosgcampos replied:
I don't remember, I copied this from the GTK pasteboard implementation. I guess we could check for other types too.

mcatanzaro replied:
https://bugs.webkit.org/show_bug.cgi?id=218519#c33

mcatanzaro commented on Source/WebKit/UIProcess/wpe/WebPasteboardProxyWPE.cpp:
```diff
@@ -0,0 +1,155 @@
+        completionHandler({ });
+        return;
+    }
+#endif
```
So I understand the intent here is "do not execute this code if the new WPE Platform API is actually being used.

But two problems. First, this is not obvious. You can make it obvious and self-documenting by adding a simple function:

```
bool isWPEPlatformAPIEnabled()
{
#if ENABLE(WPE_PLATFORM)
    return g_type_class_peek(WPE_TYPE_DISPLAY);
#else
    return false;
#endif
}
```

OK, so that's better, but second problem is: why bail early when the new API is enabled? The pasteboard classes are too complicated and it's hard for me to understand. Is this future work, or is WebPasteboardProxy not expected to be used for some reason...?

carlosgcampos replied:
Yes, this is in preparation for the clipboard support with the new api. We currently crash because we try to use libwpe, so for now the early return fixes the crashes, but it will include the actual implementation in a follow up patch.

carlosgcampos committed:
[WPE] Use new Pasteboard API
https://bugs.webkit.org/show_bug.cgi?id=291935

Reviewed by Michael Catanzaro.

Switch to use the new Pasteboard API, reusing GTK implementation when
possible. This also fixes a crash with the new API when the pasteboard
is used, since it uses libwpe unconditionally. New Pasteboard API allows
to enable several features that were disabled for WPE, but they won't
work with the old API because libwpe only supports text items in clipboard.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/editing/WebContentReader.cpp:
* Source/WebCore/editing/WebContentReader.h:
* Source/WebCore/editing/glib/EditorGLib.cpp: Renamed from Source/WebCore/editing/gtk/EditorGtk.cpp.
(WebCore::Editor::pasteWithPasteboard):
(WebCore::Editor::platformCopyFont):
(WebCore::Editor::platformPasteFont):
(WebCore::elementURL):
(WebCore::getImageForElement):
(WebCore::Editor::writeImageToPasteboard):
(WebCore::Editor::writeSelectionToPasteboard):
(WebCore::Editor::webContentFromPasteboard):
* Source/WebCore/editing/glib/WebContentReaderGLib.cpp: Renamed from Source/WebCore/editing/gtk/WebContentReaderGtk.cpp.
(WebCore::WebContentReader::readFilePath):
(WebCore::WebContentReader::readHTML):
(WebCore::WebContentReader::readPlainText):
(WebCore::WebContentReader::readImage):
(WebCore::WebContentReader::readURL):
(WebCore::shouldReplaceSubresourceURL):
(WebCore::WebContentMarkupReader::readHTML):
* Source/WebCore/platform/Pasteboard.h:
* Source/WebCore/platform/PasteboardCustomData.h:
(WebCore::PasteboardCustomData::wpeType):
* Source/WebCore/platform/PasteboardStrategy.h:
* Source/WebCore/platform/PlatformPasteboard.h:
* Source/WebCore/platform/SourcesGLib.txt:
* Source/WebCore/platform/glib/SelectionData.cpp: Renamed from Source/WebCore/platform/gtk/SelectionData.cpp.
(WebCore::SelectionData::SelectionData):
(WebCore::replaceNonBreakingSpaceWithSpace):
(WebCore::SelectionData::setText):
(WebCore::SelectionData::setURIList):
(WebCore::SelectionData::setURL):
(WebCore::SelectionData::urlLabel const):
(WebCore::SelectionData::clearAllExceptFilenames):
(WebCore::SelectionData::clearAll):
* Source/WebCore/platform/glib/SelectionData.h: Renamed from Source/WebCore/platform/gtk/SelectionData.h.
(WebCore::SelectionData::text const):
(WebCore::SelectionData::hasText const):
(WebCore::SelectionData::clearText):
(WebCore::SelectionData::setMarkup):
(WebCore::SelectionData::markup const):
(WebCore::SelectionData::hasMarkup const):
(WebCore::SelectionData::clearMarkup):
(WebCore::SelectionData::url const):
(WebCore::SelectionData::hasURL const):
(WebCore::SelectionData::clearURL):
(WebCore::SelectionData::uriList const):
(WebCore::SelectionData::filenames const):
(WebCore::SelectionData::hasURIList const):
(WebCore::SelectionData::hasFilenames const):
(WebCore::SelectionData::clearURIList):
(WebCore::SelectionData::setImage):
(WebCore::SelectionData::image const):
(WebCore::SelectionData::hasImage const):
(WebCore::SelectionData::clearImage):
(WebCore::SelectionData::setCanSmartReplace):
(WebCore::SelectionData::canSmartReplace const):
(WebCore::SelectionData::addBuffer):
(WebCore::SelectionData::buffers const):
(WebCore::SelectionData::buffer):
(WebCore::SelectionData::clearBuffers):
(WebCore::SelectionData::setCustomData):
(WebCore::SelectionData::customData const):
(WebCore::SelectionData::hasCustomData const):
(WebCore::SelectionData::clearCustomData):
* Source/WebCore/platform/libwpe/PlatformPasteboardLibWPE.cpp:
(WebCore::PlatformPasteboard::changeCount const):
(WebCore::PlatformPasteboard::write):
* Source/WebCore/platform/wpe/PasteboardWPE.cpp: Added.
(WebCore::Pasteboard::createForCopyAndPaste):
(WebCore::Pasteboard::Pasteboard):
(WebCore::Pasteboard::writeString):
(WebCore::Pasteboard::writePlainText):
(WebCore::Pasteboard::write):
(WebCore::Pasteboard::writeTrustworthyWebURLsPboardType):
(WebCore::Pasteboard::clear):
(WebCore::Pasteboard::canSmartReplace):
(WebCore::Pasteboard::read):
(WebCore::Pasteboard::hasData):
(WebCore::Pasteboard::typesSafeForBindings):
(WebCore::Pasteboard::typesForLegacyUnsafeBindings):
(WebCore::Pasteboard::readOrigin):
(WebCore::Pasteboard::readString):
(WebCore::Pasteboard::readStringInCustomData):
(WebCore::Pasteboard::fileContentState):
(WebCore::Pasteboard::writeMarkup):
(WebCore::Pasteboard::writeCustomData):
(WebCore::Pasteboard::changeCount const):
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Scripts/webkit/messages.py:
(conditions_for_header):
* Source/WebKit/Shared/Pasteboard.serialization.in:
* Source/WebKit/Shared/glib/SelectionData.serialization.in: Renamed from Source/WebKit/Shared/gtk/ArgumentCodersGtk.serialization.in.
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/WebPasteboardProxy.cpp:
* Source/WebKit/UIProcess/WebPasteboardProxy.h:
* Source/WebKit/UIProcess/WebPasteboardProxy.messages.in:
* Source/WebKit/UIProcess/wpe/WebPasteboardProxyWPE.cpp: Added.
(WebKit::WebPasteboardProxy::getTypes):
(WebKit::WebPasteboardProxy::readText):
(WebKit::WebPasteboardProxy::readFilePaths):
(WebKit::WebPasteboardProxy::readBuffer):
(WebKit::WebPasteboardProxy::writeToClipboard):
(WebKit::WebPasteboardProxy::clearClipboard):
(WebKit::WebPasteboardProxy::typesSafeForDOMToReadAndWrite):
(WebKit::WebPasteboardProxy::writeCustomData):
(WebKit::WebPasteboardProxy::allPasteboardItemInfo):
(WebKit::WebPasteboardProxy::informationForItemAtIndex):
(WebKit::WebPasteboardProxy::getPasteboardItemsCount):
(WebKit::WebPasteboardProxy::readURLFromPasteboard):
(WebKit::WebPasteboardProxy::readBufferFromPasteboard):
(WebKit::WebPasteboardProxy::getPasteboardChangeCount):
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.h:

Canonical link: https://commits.webkit.org/294103@main
